Vanish

applenick opened the pull request:
# Vanish 👻 
This PR implements a vanish mode for staff members, allowing them to appear offline to regular players. 

When a player is vanished the following occurs:
1. A leave message is broadcasted
2. Forced to observing team, and will be unable to join using picker or command
3. Visibility is reset so only other staff can see you
4. Tab name is completely removed for regular players (staff remain visible)

**Additional notes**
* Vanish status persists until the server restarts
* A broadcast is sent to admin chat when a vanished staff member actually joins/quits
* Attempting to /msg <vanished player> will send an user not found response
* All global/team/dms are disabled for vanished players (only admin chat)
* Vanished players are excluded from `/list`

## Commands
`/vanish` - Will vanish you from the server, making it appear like you are offline. 

`/unvanish` - Will make you visible again to players, will seem as if you join rejoined.

`/list` - Displays a list of online players (now grouped by teams or participants)


## Screenshots

First is an example of using `/vanish` from the perspective both staff member and regular.

Staff:
![staff-view](https://user-images.githubusercontent.com/3377659/80566903-dd633980-89a8-11ea-809a-4c3e5a31fc45.gif)

Regular:
![player-view](https://user-images.githubusercontent.com/3377659/80566914-e48a4780-89a8-11ea-85af-db83d2ffa612.gif)

As you can see the staff member is sent a message along with refreshing their observer kit to remove the picker. From the regular view, it appears as if the staff member has quit the server.

Below is the broadcast message sent to admin chat when a vanished staff member joins/quits. 
![Screen Shot 2020-04-28 at 11 27 47 PM](https://user-images.githubusercontent.com/3377659/80567032-22876b80-89a9-11ea-9d0f-31eb89ef59d4.png)

The final set are of the new `/list` command, as you will see it is neat and organized by teams. Also included an example view of when a staff member views the list, as staff will be able to see players who are vanished as identified with a strikethrough (same as the tab list). 
![Screen Shot 2020-04-28 at 11 20 46 PM](https://user-images.githubusercontent.com/3377659/80567170-75f9b980-89a9-11ea-9599-7519e579e5bb.png)

![Screen Shot 2020-04-28 at 5 15 36 PM](https://user-images.githubusercontent.com/3377659/80567181-7abe6d80-89a9-11ea-9605-58b6566326db.png)

![Screen Shot 2020-04-28 at 5 15 26 PM](https://user-images.githubusercontent.com/3377659/80567188-7eea8b00-89a9-11ea-8dd4-843e7102c741.png)
## Final Notes
I believe this PR handles most of the features related to vanish, however there are a few accommodations that will have to be made for some setups. 

For example on OCC, observers have access to the commandbook `/tp` command. Even if a player is vanished, if a non-staff performs `/tp <vanished player>` they will still be teleported to them. 

I plan on forking CommandBook and adding in checks that will accommodate this, however it's always possible other servers may use different plugins for various things. I’ll link my CommandBook fork here once it is ready, for future reference. 

That is something that will have to be taken into account when running PGM and using the vanished feature. I did include both a `PlayerVanishEvent` and `isVanished` metadata that is attached to the player. So there are multiple ways one could check for vanished players via separate plugins. 

As always feedback is appreciated, and I will make adjustments where necessary :thumbsup:

Signed-off-by: applenick <applenick@users.noreply.github.com>

KingOfSquares commented on core/src/main/java/tc/oc/pgm/commands/UserCommands.java:
```diff
@@ -0,0 +1,112 @@
+import tc.oc.pgm.util.component.types.PersonalizedTranslatable;
+import tc.oc.pgm.util.named.NameStyle;
+
+public class UserCommands {
```
`UserCommands` feels so generic many other commands could fit in here, what do you think about `ListCommands`?

KingOfSquares commented on core/src/main/java/tc/oc/pgm/community/events/PlayerVanishEvent.java:
```diff
@@ -0,0 +1,35 @@
+
+public class PlayerVanishEvent extends ExtendedCancellable {
+
+  private MatchPlayer vanisher;
+  private boolean vanished;
```
Both of these can be `final`?

Electroid replied:
Yes

KingOfSquares commented on core/src/main/java/tc/oc/pgm/community/features/VanishManager.java:
```diff
@@ -0,0 +1,176 @@
+  private static final String VANISH_KEY = "isVanished";
+  private static final MetadataValue VANISH_VALUE = new FixedMetadataValue(PGM.get(), true);
+
+  private List<UUID> vanishedPlayers;
```
`private final List<UUID>` ?

KingOfSquares commented on core/src/main/java/tc/oc/pgm/listeners/ChatDispatcher.java:
```diff
@@ -167,12 +170,27 @@ public void sendAdmin(Match match, MatchPlayer sender, @Nullable @Text String me
     MatchPlayer matchReceiver = manager.getPlayer(receiver);
     if (matchReceiver != null) {
+
+      // Vanish Check - Don't allow messages to vanished
+      if (vanish.isVanished(matchReceiver.getId())) {
+        sender.sendMessage(
+            new PersonalizedTranslatable("command.playerNotFound")
+                .getPersonalizedText()
+                .color(ChatColor.RED));
+        return;
+      }
```
Exclude staff from this? As in staff can send vanished staff messages? Maybe its enough with adminchat, what do you think?

applenick replied:
Will there be any reason to message other staff members? Using `/a` should suffice in my opinion and most staff usually know each other outside of the server, so any private conversations can be held via other platforms.

KingOfSquares commented on core/src/main/java/tc/oc/pgm/api/PGM.java:
```diff
@@ -35,6 +36,8 @@
   ScheduledExecutorService getAsyncExecutor();
 
+  VanishManager getVanishManager();
+
```
This makes PGM depend on the vanish feature being present, which goes against #326 since it currently is a community feature. Maybe check for the vanish value in the same way you intended for the util package? (See other comment for where i'm thinking about)

Electroid replied:
Hmm, I'd say make `VanishManager` an interface in `tc.oc.pgm.api`. With 2 implementation, a `NoopVanishManager` in `pgm` and a `VanishManagerImpl` in `community`.

KingOfSquares commented on core/src/main/java/tc/oc/pgm/community/features/VanishManager.java:
```diff
@@ -0,0 +1,176 @@
+              player.getBukkit()));
+    }
+
+    // Keep track of the UUID and apply/remove META data, so we can detect vanish status from other
+    // projects (i.e utils)
+    if (vanish) {
+      addVanished(player.getId());
+      player.getBukkit().setMetadata(VANISH_KEY, VANISH_VALUE);
+      player.getSettings().setValue(SettingKey.CHAT, SettingValue.CHAT_ADMIN);
+    } else {
+      removeVanished(player.getId());
+      player.getBukkit().removeMetadata(VANISH_KEY, VANISH_VALUE.getOwningPlugin());
+    }
```
See other comment about #326

EudaldC commented on core/src/main/java/tc/oc/pgm/community/features/VanishManager.java:
```diff
@@ -0,0 +1,176 @@
+  @Command(
+      aliases = {"vanish", "disappear", "v"},
+      desc = "Vanish from the server",
+      perms = Permissions.STAFF)
```
I really dislike the approach of pgm.staff having all the perms, as it makes the plugin permissions a lot less customizable. I may want some staff groups in Stratus to be able to vanish but not freeze or lookup alts, especially with permission realms in mind, imagine a tournament setting for example.

I know it's down to @Electroid to decide in which parts of the project complexity is to be avoided, but I strongly believe that simplifying permission nodes is not worth the result.

Electroid replied:
I think `pgm.vanish` should be ok.

Pablete1234 replied:
The issue if you make it `pgm.vanish` is that to setup permission for your staff you start getting a long list that becomes hard to manage. You end up doing permission node hunting to find out what to give and what not to give your mods access to, happened on 1.12 pgm and it became kinda hard to manage

Pablete1234 commented:
You could probably display in the action bar a message that reminds you of your vanished status

Electroid commented on core/src/main/java/tc/oc/pgm/community/events/PlayerVanishEvent.java:
```diff
@@ -0,0 +1,35 @@
+import tc.oc.pgm.api.event.ExtendedCancellable;
+import tc.oc.pgm.api.player.MatchPlayer;
+
+public class PlayerVanishEvent extends ExtendedCancellable {
```
Make sure to extend existing events, seems like this would fall under `MatchPlayerEvent`

applenick commented:
Addressed most of the issues brought up from the feedback.

I decided to give separate permissions for both vanish and freeze (`pgm.vanish` `pgm.freeze`), though I do agree with @Pablete1234’s point. It may be better longterm to switch moderation and similar commands to a `pgm.staff.<type>` format. 

Let me know if there is anything else I can adjust😄

Electroid commented on core/src/main/java/tc/oc/pgm/api/player/VanishManager.java:
```diff
@@ -0,0 +1,37 @@
+   *
+   * @return A list of {@link MatchPlayer}
+   */
+  List<MatchPlayer> getOnlineVanished();
```
Unless it needs to be a list. `Collection<MatchPlayer> getVanished()`

Electroid commented on core/src/main/java/tc/oc/pgm/api/player/VanishManager.java:
```diff
@@ -0,0 +1,37 @@
+   * @param quiet - Whether to broadcast related join/quit messages
+   * @throws CommandException When player is already vanished/unvanished
+   */
+  void setVanished(MatchPlayer player, boolean vanish, boolean quiet) throws CommandException;
```
We shouldn't use `CommandException` here, that's a concern of the commands not the interface.

How about this method return a `boolean`?

Electroid commented on core/src/main/java/tc/oc/pgm/commands/JoinCommands.java:
```diff
@@ -47,7 +47,7 @@ public static void join(
     }
 
-    if (sender.hasPermission(Permissions.JOIN_FORCE) && force) {
+    if (sender.hasPermission(Permissions.JOIN_FORCE) && force && !player.isVanished()) {
```
Does this check need to be inside `JoinMatchModule`? I think so, because it should also guard against the picker.

Electroid commented on core/src/main/java/tc/oc/pgm/commands/ListCommands.java:
```diff
@@ -0,0 +1,112 @@
+    } else {
+      sendTeamInfo(
+          sender,
+          new PersonalizedTranslatable("commands.users.participants"),
```
It would be great if we can start using `TextComponent` and `TranslatableComponent` with the new text library. Soon I won't be accepting any more changes with the old text. Or else we'll never migrate.

applenick replied:
I was wondering about this, as I seem to have run into a few issues. First using `MatchPlayer#showHotbar` did not actually show the hotbar. After searching the repo, I was unable to find any calls to the new hotbar, so not sure if that’s been tested. 

Also in `ListCommands` I’m utilizing `Components#join` to neatly format the player names. Since that’s still using the old component system, what should be done there? I can migrate the method if you like, however there are quite a few calls to that method from `ModerationCommands`.

 I’m planning on a PR soon that will have some changes to `ModerationCommands`, maybe it would be more appropriate to migrate all the moderations commands to the new text library then? 

I did however use `TranslatableComponent` for the main vanish commands. 👍

Electroid replied:
Great, if you have some blockers don’t worry about it for now.

Electroid commented on core/src/main/java/tc/oc/pgm/community/events/PlayerVanishEvent.java:
```diff
@@ -0,0 +1,30 @@
+
+public class PlayerVanishEvent extends MatchPlayerEvent {
+
+  private final boolean vanished;
```
Since this is sorted in the `MatchPlayer` object, you don't need to store it in the event as well.

Electroid commented on core/src/main/java/tc/oc/pgm/community/events/PlayerVanishEvent.java:
```diff
@@ -0,0 +1,30 @@
+  }
+
+  public boolean isVanished() {
+    return vanished;
```
`vanisher.isVanished()`

Electroid commented on core/src/main/java/tc/oc/pgm/community/features/VanishManagerImpl.java:
```diff
@@ -0,0 +1,209 @@
+    // projects (i.e utils)
+    if (vanish) {
+      addVanished(player.getId());
+      player.getBukkit().setMetadata(VANISH_KEY, VANISH_VALUE);
+      player.getSettings().setValue(SettingKey.CHAT, SettingValue.CHAT_ADMIN);
```
Seems like `addVanished` should do this?

Electroid commented on core/src/main/java/tc/oc/pgm/community/features/VanishManagerImpl.java:
```diff
@@ -0,0 +1,209 @@
+      player.getSettings().setValue(SettingKey.CHAT, SettingValue.CHAT_ADMIN);
+    } else {
+      removeVanished(player.getId());
+      player.getBukkit().removeMetadata(VANISH_KEY, VANISH_VALUE.getOwningPlugin());
```
Same here

Electroid commented on util/src/main/i18n/templates/strings.properties:
```diff
@@ -1041,6 +1042,25 @@ stats.current = Current stats
 stats.disabled = You have disabled stat tracking
 
+vanish.activate = You have vanished!
+vanish.deactivate = You are now visible!
```
I think you want to use the same vanish terminology, not visible.

Electroid commented:
Merged into #431
